packages/bpk-component-calendar/src/date-utils.ts

```diff
@@ -110,7 +115,12 @@ function getCalendarMonthWeeks(date: Date, weekStartsOn: number) {
   for (let i = 0; i < 6; i += 1) {
     const currWeek = [];
     for (let j = 0; j < 7; j += 1) {
-      currWeek.push(dateAtStartOfDay(year, month, day));
+      const currDate = dateAtStartOfDay(year, month, day);
+      currWeek.push({
+        val: currDate,
+        textLabel: formatDate ? formatDate(currDate) : 'loading...',
```

Comment:
Is this rendered to the user? We can't have `en` strings just output. I'm wondering if this api should just be async & return a `Promise<DateProps[][]>`?

Reply:
It's added as aria-labels but replaced with the real value within a second or so - would an empty string be better?

Reply:
Everything other than the textLabel is required for initial render, so simplifying the API in the way you suggest would re-introduce the performance bottleneck of this date formatting operation. I can split the method in two for clarity sake?

Reply:
> performance bottleneck

Why would it reintroduce a performance bottleneck? The UI experience might be different (the user might have to wait for the formatting to occur), but there should be no main thread blocking etc. 

---

Thinking about this a little more though, isn't this function only being run like 42 * 12 (~500) times? Roughly for rendering an entire year. With the doc you linked in mind this seemed to take around 500ms - is that right? That seems, well, very very slow. You could go to a server and fetch the formatting information at that speed. Is the problem not that the formatting function itself is incredibly inefficient, should we focus on that? Like if the format fn is `.toString()` do we see big improvements to INP too?

Reply:
The performance bottleneck exists because the component cannot currently instantiate (and subsequently render) the calendar component until all dates have been formatted. The blocking time this occupies is shown in the profiler screenshots within the linked document. 

- The function is initially run for 4 months of dates, with more added on scroll (a strategy to reduce INP). This is synchronous and whilst relatively fast on Desktop, it is much slower on Mobile devices. 
- It is not 500ms for the average user, these are stats taken from my machine with a synthetic 6x CPU slowdown. The important thing to note is the ratios, 22% reduction in INP by removing unnecessary / duplicative date format calls and a further 21% reduction in INP (43% total) by using requestIdleCallback. Numbers are a guideline but does indicate this is an effective strategy
- Skyscanner are mandated to use saddlebag-localisation for locale-based date formatting. This uses Globalize under the hood meaning it is not an option for us to address the inefficiency of that function.
- A server solution may not satisfy Skyscanner date formatting guidelines and would be more prone to failure. It would still require deferred addition of the aria-labels as this solution offers.